### DISCUSSION TITLE
Script vs functions??

### DISCUSSION THREAD

**Discussion by BarbzYHOOL** [fish-shell/fish-shell#10118] [Q&A] [answered]
Hello,
I made a script for manipulating audio files, I placed it in my local bin dir and I execute it, it's good
Now I need an argument for the script so I was like "why not make it into a fish function inside the fish functions directory?"
And then I'm like "yeah, why not make functions instead of scripts?"
and I'm confused, what is the best, why and when?

**Comment by faho** [comment, marked as answer]
A function will run directly inside of your shell.

That means it'll be able to touch your shell's environment directly - it can set variables that stay, it can `cd`, it can run other functions, ...

A script will run in its own shell. That means it's somewhat isolated - it won't be affected by many things in your shell (except for exported variables), and it can't do many things to your shell. Fish is a bit special in this regard because it reads all config by default (i.e. unless you run it specifically with `--no-config`), so even a script will have access to your functions etc, so the difference is less than it would be in other shells.

If you want to give your thing to people who use other shells, it'll have to be a script.

**Reply by BarbzYHOOL** [reply]
In which other cases should I use a script over a function and vice versa?
